src/access/universal.test.ts

```diff
@@ -0,0 +1,889 @@
+  });
+
+  it("returns the new effective access for ACP-controlled Resources", async () => {
+    const getResourceInfoWithAcr = jest.spyOn(
+      acpLowLevel,
+      "getResourceInfoWithAcr"
+    );
+    const mockedResource = mockSolidDatasetFrom(
+      "https://arbitrary.pod/resource"
+    );
+    const mockedAcr = mockAcrFor("https://arbitrary.pod/resource");
+    const mockedResourceWithAcr = addMockAcrTo(mockedResource, mockedAcr);
+    getResourceInfoWithAcr.mockResolvedValueOnce(mockedResourceWithAcr);
+    jest
+      .spyOn(acpModule, "internal_setAgentAccess")
+      .mockReturnValueOnce(mockedResourceWithAcr);
+    jest
+      .spyOn(acpLowLevel, "saveAcrFor")
+      .mockResolvedValueOnce(undefined as any);
+    jest
+      .spyOn(acpModule, "internal_getAgentAccess")
+      .mockReturnValueOnce({ append: true });
+
+    const returnedAccess = await setAgentAccess(
+      "https://arbitrary.pod/resource",
+      "https://arbitrary.pod/profile#me",
+      { read: true }
+    );
+
+    expect(returnedAccess).toStrictEqual({ append: true });
+  });
+
+  it("calls out to the well-tested WAC API for Resources with an ACR", async () => {
+    const getResourceInfoWithAcr = jest.spyOn(
+      acpLowLevel,
+      "getResourceInfoWithAcr"
+    );
+    const mockedResource = mockSolidDatasetFrom(
+      "https://arbitrary.pod/resource"
+    );
+    const mockedResourceWithAcl = addMockResourceAclTo(mockedResource);
+    getResourceInfoWithAcr.mockResolvedValueOnce(mockedResourceWithAcl as any);
+    const wacSetAgentAccess = jest.spyOn(wacModule, "setAgentResourceAccess");
+    wacSetAgentAccess.mockResolvedValue(null);
+
+    await setAgentAccess(
+      "https://arbitrary.pod/resource",
+      "https://arbitrary.pod/profile#me",
+      { read: true }
+    );
+
+    expect(wacSetAgentAccess).toHaveBeenCalledTimes(1);
+  });
+
+  it("throws an error when setting differing controlRead and controlWrite for a WAC-controlled Resource", async () => {
+    const getResourceInfoWithAcr = jest.spyOn(
+      acpLowLevel,
+      "getResourceInfoWithAcr"
+    );
+    const mockedResource = mockSolidDatasetFrom("https://some.pod/resource");
+    const mockedResourceWithAcl = addMockResourceAclTo(mockedResource);
+    getResourceInfoWithAcr.mockResolvedValueOnce(mockedResourceWithAcl as any);
+    const wacSetAgentAccess = jest.spyOn(wacModule, "setAgentResourceAccess");
+    wacSetAgentAccess.mockResolvedValue(null);
+
+    await expect(
+      setAgentAccess(
+        "https://some.pod/resource",
+        "https://arbitrary.pod/profile#me",
+        { controlRead: true, controlWrite: undefined }
+      )
+    ).rejects.toThrow(
+      "When setting access for a Resource in a Pod implementing Web Access Control (i.e. [https://some.pod/resource]), `controlRead` and `controlWrite` should have the same value."
+    );
+  });
+
+  it("returns null when the ACP module could not update the ACR", async () => {
```

Comment:
Idem, should this also be tested for ACL ?

Reply:
Yep, will do.

Reply:
b82640fe